pkg/kubelet/cm/dra/state/checkpoint.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"encoding/json"
+
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager"
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager/checksum"
+)
+
+var _ checkpointmanager.Checkpoint = &DRAManagerCheckpoint{}
+
+// List of claim info to store in checkpoint
+type ClaimInfoStateList []*ClaimInfoState
+
+const checkpointSchemeVersion = "v1"
+
+// DRAManagerCheckpoint struct is used to store pod dynamic resources assignments in a checkpoint
+type DRAManagerCheckpoint struct {
+	Version  string             `json:"version"`
+	Entries  ClaimInfoStateList `json:"entries,omitempty"`
+	Checksum checksum.Checksum  `json:"checksum"`
+}
+
+// NewDRAManagerCheckpoint returns an instance of Checkpoint
+func NewDRAManagerCheckpoint() *DRAManagerCheckpoint {
+	//nolint:staticcheck // unexported-type-in-api user-facing error message
+	return &DRAManagerCheckpoint{
+		Version: checkpointSchemeVersion,
+		Entries: ClaimInfoStateList{},
+	}
+}
+
+// MarshalCheckpoint returns marshalled checkpoint
+func (dc *DRAManagerCheckpoint) MarshalCheckpoint() ([]byte, error) {
+	// make sure checksum wasn't set before so it doesn't affect output checksum
+	dc.Checksum = 0
+	dc.Checksum = checksum.New(dc)
+	return json.Marshal(*dc)
+}
+
+// UnmarshalCheckpoint tries to unmarshal passed bytes to checkpoint
+func (dc *DRAManagerCheckpoint) UnmarshalCheckpoint(blob []byte) error {
+	return json.Unmarshal(blob, dc)
+}
+
+// VerifyChecksum verifies that current checksum of checkpoint is valid
+func (dc *DRAManagerCheckpoint) VerifyChecksum() error {
+	ck := dc.Checksum
+	dc.Checksum = 0
+	err := ck.Verify(dc)
```

Comment:
@moshe010 This looks like we're always verifying checksum 0. Am I missing something?

Reply:
yes we don't want that the checksum will effect the output checksum  see https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/memorymanager/state/checkpoint.go#L46-L51

Reply:
@moshe010 I understand why marshalling code does that. My question is about verifying code. What's the point of verifying 0 checksum only?

Reply:
we don't verify  0 checksum. we verify all the DRAManagerCheckpoint struct ( I am passing dc)   but before doing that we put checksum =0 because the struct include the checksum.  This is how we generated the checksum in the marshalling code so when we verify we passing the DRAManagerCheckpoint  struct with checksum =0.

Reply:
ah, ok. now I understand. Thanks for the explanations!